File: backend/app/api/v1/monitor.py
from fastapi import APIRouter, HTTPException, status

from app.core.responses import ok
from app.services import monitor_service

router = APIRouter(prefix="/monitor", tags=["monitor"])


@router.get("/health")
def get_health():
    return ok(monitor_service.get_health())


@router.get("/metrics/nodes")
def get_node_metrics(
    nodeIds: list[str] | None = None,
    startTime: int | None = None,
    endTime: int | None = None,
    interval: int | None = None,
):
    result = monitor_service.get_node_metrics(nodeIds, startTime, endTime, interval)
    return ok(result)


@router.get("/metrics/links")
def get_link_metrics(
    linkIds: list[str] | None = None,
    startTime: int | None = None,
    endTime: int | None = None,
    interval: int | None = None,
):
    result = monitor_service.get_link_metrics(linkIds, startTime, endTime, interval)
    return ok(result)


@router.get("/alerts")
def get_alerts(acknowledged: bool | None = None, level: str | None = None):
    result = monitor_service.get_alerts(acknowledged=acknowledged, level=level)
    return ok(result)


@router.post("/alerts/{id}/acknowledge")
def acknowledge_alert(id: str):
    if not monitor_service.acknowledge_alert(id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="告警不存在")
    return ok(None)


@router.get("/alert-rules")
def get_alert_rules():
    result = monitor_service.get_alert_rules()
    return ok(result)


@router.post("/alert-rules")
def create_alert_rule(body: dict):
    name = body.get("name", "")
    enabled = body.get("enabled", True)
    type_ = body.get("type", "")
    threshold = body.get("threshold", 0.0)
    duration = body.get("duration", 0)
    actions = body.get("actions", [])
    rule = monitor_service.create_alert_rule(name, enabled, type_, threshold, duration, actions)
    return ok(rule)


@router.put("/alert-rules/{id}")
def update_alert_rule(id: str, body: dict):
    rule = monitor_service.update_alert_rule(
        id,
        name=body.get("name"),
        enabled=body.get("enabled"),
        type=body.get("type"),
        threshold=body.get("threshold"),
        duration=body.get("duration"),
        actions=body.get("actions"),
    )
    if not rule:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="告警规则不存在")
    return ok(rule)


@router.delete("/alert-rules/{id}")
def delete_alert_rule(id: str):
    if not monitor_service.delete_alert_rule(id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="告警规则不存在")
    return ok(None)


@router.get("/terminal/logs")
def get_terminal_logs(nodeId: str | None = None, lines: int = 100):
    result = monitor_service.get_terminal_logs(nodeId, lines)
    return ok(result)

File: frontend/src/api/monitor.ts
import axios from './axios'
import type { ApiResponse } from './axios'
import type { 
  NodeMetrics, 
  LinkMetrics, 
  SystemHealth, 
  Alert, 
  AlertRule 
} from '@/model/monitor'

export interface MetricsQuery {
  nodeIds?: string[]
  linkIds?: string[]
  startTime?: number
  endTime?: number
  interval?: number // seconds
}

export const monitorApi = {
  // 获取系统健康状态
  getHealth: () => {
    return axios.get<void, ApiResponse<SystemHealth>>('/monitor/health')
  },

  // 获取节点指标
  getNodeMetrics: (params: MetricsQuery) => {
    return axios.get<MetricsQuery, ApiResponse<NodeMetrics[]>>(
      '/monitor/metrics/nodes',
      { params }
    )
  },

  // 获取链路指标
  getLinkMetrics: (params: MetricsQuery) => {
    return axios.get<MetricsQuery, ApiResponse<LinkMetrics[]>>(
      '/monitor/metrics/links',
      { params }
    )
  },

  // 获取告警列表
  getAlerts: (params?: { acknowledged?: boolean; level?: string }) => {
    return axios.get<typeof params, ApiResponse<Alert[]>>('/monitor/alerts', { params })
  },

  // 确认告警
  acknowledgeAlert: (id: string) => {
    return axios.post<void, ApiResponse<void>>(`/monitor/alerts/${id}/acknowledge`)
  },

  // 获取告警规则
  getAlertRules: () => {
    return axios.get<void, ApiResponse<AlertRule[]>>('/monitor/alert-rules')
  },

  // 创建告警规则
  createAlertRule: (data: Omit<AlertRule, 'id'>) => {
    return axios.post<typeof data, ApiResponse<AlertRule>>('/monitor/alert-rules', data)
  },

  // 更新告警规则
  updateAlertRule: (id: string, data: Partial<AlertRule>) => {
    return axios.put<typeof data, ApiResponse<AlertRule>>(`/monitor/alert-rules/${id}`, data)
  },

  // 删除告警规则
  deleteAlertRule: (id: string) => {
    return axios.delete<void, ApiResponse<void>>(`/monitor/alert-rules/${id}`)
  },

  // 获取终端日志（历史片段，用于非 WebSocket 场景）
  getTerminalLogs: (params?: { nodeId?: string; lines?: number }) => {
    return axios.get<typeof params, ApiResponse<string>>('/monitor/terminal/logs', { params })
  },
}
